src/main/java/io/leonis/subra/ipc/serialization/protobuf/vision/GeometryDeducer.java

```diff
@@ -0,0 +1,121 @@
+package io.leonis.subra.ipc.serialization.protobuf.vision;
+
+import io.leonis.algieba.geometry.CardinalDirection;
+import io.leonis.subra.game.data.*;
+import io.leonis.subra.ipc.serialization.protobuf.vision.GeometryDeducer.Geometry;
+import io.leonis.zosma.game.engine.Deducer;
+import java.util.Set;
+import java.util.stream.*;
+import lombok.Value;
+import org.nd4j.linalg.factory.Nd4j;
+import org.reactivestreams.Publisher;
+import org.robocup.ssl.Geometry.*;
+import org.robocup.ssl.Wrapper.WrapperPacket;
+import reactor.core.publisher.Flux;
+
+/**
+ * The Class GeometryDeducer.
+ *
+ * This class represents a {@link Deducer} of {@link GeometryData} from {@link WrapperPacket}.
+ *
+ * @author Rimon Oz
+ */
+public class GeometryDeducer implements Deducer<WrapperPacket, Geometry> {
+  @Override
+  public Publisher<Geometry> apply(final Publisher<WrapperPacket> detectionFramePublisher) {
+    return Flux.from(detectionFramePublisher)
+        .filter(WrapperPacket::hasGeometry)
+        .map(WrapperPacket::getGeometry)
+        .map(geom ->
+            new Geometry(
+                new Field.State(
+                    geom.getField().getFieldWidth(),
+                    geom.getField().getFieldLength(),
+                    geom.getField().getFieldLinesList().stream()
+                        .map(fieldLineSegment ->
+                            new FieldLine.State(
+                                fieldLineSegment.getP1().getX(),
+                                fieldLineSegment.getP1().getY(),
+                                fieldLineSegment.getP2().getX(),
+                                fieldLineSegment.getP2().getY(),
+                                fieldLineSegment.getThickness()))
+                        .collect(Collectors.toSet()),
+                    geom.getField().getFieldArcsList().stream()
+                        .map(fieldCircularArc ->
+                            new FieldArc.State(
+                                fieldCircularArc.getCenter().getX(),
+                                fieldCircularArc.getCenter().getY(),
+                                fieldCircularArc.getA1(),
+                                fieldCircularArc.getA2(),
+                                fieldCircularArc.getThickness(),
+                                fieldCircularArc.getRadius()))
+                        .collect(Collectors.toSet())),
+                Stream.of(geom)
+                    .flatMap(packet -> Stream.of(
+                        this.createGoal(TeamColor.BLUE, CardinalDirection.NORTH, packet.getField()),
```

Comment:
Let's expose the coupling between Cardinal direction and teamcolor

Reply:
We might need to solve this in a different PR. I can see how we would want to publish that coupling and introduce a couple of `Suppliers` and a container for these kinds of settings (e.g. `GameSetup implements TeamDirection.SetSupplier`, with `TeamDirection` being a simple
```java
interface TeamDirection {
  TeamIdentifier getTeamIdentifier();
  CardinalDirection getDirection();

  interface SetSupplier {
     Set<TeamDirection> getTeamDirections();
  }
}
```

Which can be simplified if we want to keep support to just 2 teams of course.

Reply:
see https://github.com/delta-leonis/subra/pull/28